### ISSUE TITLE
validation accuracy superior to training accuracy

### ISSUE THREAD

**Issue by JulienDucrest** [fchollet/keras#1761]
Hi,
I'm working on a deep neural network (Sequential Dense), but i don't understand why the training accuracy is lower than the validation accuracy (the two sets being separated and having roughly the same distribution).
I started working on machine learning not so long ago and i was told that a way(the only one i know) to check if my network is not overfitting is to compare validation and train accuracy. If validation accuracy start dropping while the training accuracy continue to increase that's when i should be concerned.
Problem is validation accuracy is higher than training accuracy which doesn't make any sense for me...

I'm sure i missed something somewhere (maybe the training accuracy displayed by keras is not the one i think of) or the way i put my validation data is not the good one... Or maybe am I just wrong all along.
So if any one could give me an advice, or knows if there is a way to plot or visualize the data training within Keras to prevent overfit that would help.

I'm talking about that kind of thing (i'm copy pasting a random epoch but all are roughly the same):
50000/50000 [==============================] - 32s - loss: 1.7436 - acc.: 0.5749 - val. loss: 1.5925 - val. acc.: 0.6434

My network parameters are the following : a sequential dense network of that kind 20000>1000>1000>1000>1000 
trained on ReLU for all except the last Softmax. The Loss is categorical crossentropy

**Comment by parag2489** [comment]
Check the following things when training any type of deep neural network:

1. the data used to calculate training accuracy is not identical to the data used to train your NN. This sounds weird, but possible in practice, especially in case of images, if you don't take care. For example, you train on random patches of images and calculate training accuracy on _random_ patches of same images. It is easy to forget that though they are same images, the patches are randomly selected.

2. More than the values of train and val accuracy, I would be concerned about what you said, "i'm copy pasting a random epoch but all are roughly the same". No, they can't be same. Accuracy at different epochs is mostly different, because network is learning so it is constantly changing its weights. If accuracy goes up then that means it is approaching the minima of the loss function.

3. I think you should be more concerned about getting a low training accuracy instead of getting a lower training accuracy than the validation accuracy.

4. Do all the sanity checks given [here](http://cs231n.github.io/neural-networks-3/#gradcheck). Read the entire article if possible, its very good.

5. Make sure you are doing pre-processing in the right manner. For example, make sure that mean over entire training data is zero. For testing data, subtract the mean vector of the training data from each instance of testing data. Don't subtract the mean of testing data from itself. Since, you wouldn't know the mean of testing data at runtime.

6. Check if your loss at the very first epoch makes sense. For example, in a 10-class classification problem, starting loss should be -ln(0.1) = 2.302 (given [here](http://cs231n.github.io/neural-networks-3/#gradcheck)). 

7. Again, from [here](http://cs231n.github.io/neural-networks-3/#gradcheck), overfit a tiny subset of data and make sure you can achieve zero cost. Full details in the link.

8. If nothing works, just train and test on the same data and see if you can get 90% + accuracy. Otherwise, examine your network more closely by looking at individual layer outputs (given in Keras FAQ) etc.

I am sure you will find the fault somewhere if you follow all these steps for debugging a neural network.

**Comment by neggert** [comment]
If your validation set is small and you're still underfitting, you can sometimes get higher validation accuracy just by random chance. Usually this will go away as you train more.

**Comment by jerheff** [comment]
I've been tracking the ratio of the training loss to the validation loss.  For me, the ratio starts off quite above 1 and slowly converges to 1 over time.

I don't know if you are using dropout, but in thinking about why my validation loss is lower than the training loss, I am considering this to be a key reason.  The training loss is based upon the specific randomly selected dropped-out network for that record, which is much less likely to be well tuned than the loss when using the full network (with adjusted weights) on the validation data.

**Comment by rpinsler** [comment]
This is also answered in the [FAQ](http://keras.io/faq/#why-is-the-training-loss-much-higher-than-the-testing-loss)

**Comment by JulienDucrest** [comment]
@rpinsler Yes sorry i didn't see it,  the use of dropout indeeds make the training accuracy lower than it should be.
And thanks to you @parag2489 for all thoses advices and the reading, helped me a lot.

**Comment by akshayk0406** [comment]
I am also facing the same issue. My problem is 3 class classification(with 29 columns) and classes are highly imbalance.

Architecture of network:-
- 4 Hidden layers with number of nodes in each layer be [columns, 1.25 * columns, 0.75 * columns, columns] where columns = 29.
- I am using relu activation for all hidden layers and for output layer i am using softmax.
- Adagrad Optimizer.
- I am optimizing for sparse_categorical_crossentropy loss.
- Dropout 0.1

 I observed few things which doesn't seem intuitive:-
i) To account for class imbalance, I am oversampling from minority classes. And I observe that more I oversample there is decrease in  training and validation accuracy. 
ii) validation accuracy is always greater than training accuracy irrespective of using oversampling or not.

Below is the code for the reference:-
```
def create_model(columns, num_class, optimizer='Adagrad', init_mode='normal', activation='relu',
			dropout_rate=0.1, weight_constraint=5):
		
		model = Sequential()
		num_nodes = [columns, int(1.25 * columns), int(0.75 * columns), columns]

		for i in range(len(num_nodes)):
			model.add(Dense(columns, input_dim = columns, init = init_mode, activation = activation, W_constraint = maxnorm(weight_constraint)))
			model.add(Dropout(dropout_rate))
		
		model.add(Dense(output_dim = num_class, activation = 'softmax')) #Output layer
		model.compile(loss='sparse_categorical_crossentropy', optimizer = optimizer, metrics=['fbeta_score'])
		return model

model = create_model(X.shape[1], len(labels))
model.fit(X_train, y_train, nb_epoch=num_epoch, batch_size=batch_size, validation_data=(X_test, y_test),callbacks=[history], verbose = 0)
```
